### ISSUE TITLE
perf(offpolicy): 降低 GPU collector 的 CPU-staged weight sync 延迟

### ISSUE THREAD

**Issue by TATP-233** [Motphys/UniLab#946]
> Work type: infra / performance  
> Area: off-policy / IPC（以 Motrix G1 FlashSAC 为复现 workload）  
> Scope: Small / implementation（单一结果、一个 PR）

## Owner summary

本 issue 只做一件事：在保留现有 CPU shared-memory weight snapshot 协议的前提下，降低 CUDA collector 应用新 actor 权重的稳态延迟，使 `training.collector_infer_device=gpu` 不再因为 weight refresh 出现约 10 ms 的周期边界停顿。

不引入 CUDA IPC、共享 CUDA tensor、NCCL 或 GPU 双缓冲；不新增 collector/learner queue、ack 协议或 runner lifecycle。预计修改 `SharedWeightSync`、off-policy collector worker 及其邻近测试，共 3–6 个文件、净手写改动不超过约 400 行。永久维护成本应限制为现有 state layout/version/lock contract 下的 device-aware apply fast path，不建立第二套同步协议。

## Problem

#687 为 off-policy collector 增加了可配置 inference device，并明确允许初版继续使用：

```text
learner device -> CPU shared memory -> collector infer device
```

在以下 workload 中：

```bash
uv run train --algo flashsac --task g1_walk_flat --sim motrix \
  training.collector_infer_device=gpu
```

collector actor 已常驻 CUDA，但每次检测到新 version 后，当前 `SharedWeightSync.read_weights_into()` 仍逐个遍历 actor `state_dict`，从 host shared memory 更新 CUDA actor。这个 G1 FlashSAC actor 有 31 个 state entry（20 个 parameter、11 个 persistent buffer），共 287,826 个 float32，约 1.098 MiB。

当前一次逻辑 refresh 的路径为：

```text
learner CUDA actor
  -> 31 次 `.cpu()` D2H
  -> flat CPU shared-memory snapshot
  -> 31 次 shared-memory slice copy
  -> 31 次同步式 H2D `copy_`
  -> `actor.load_state_dict(sd)` 再次遍历 state
```

这里的 31 是一次逻辑 weight refresh 内的 31 个 tensor copy，不是 31 次 collector/learner 同步。

## Baseline evidence

环境：NVIDIA GeForce RTX 4090、PyTorch `2.7.0+cu128`；`gpu` 解析为 `cuda`。为缩短采样使用：

```bash
uv run train --algo flashsac --task g1_walk_flat --sim motrix \
  training.collector_infer_device=gpu \
  algo.learning_starts=1 algo.max_iterations=30 \
  training.trace_enabled=true training.trace_cuda_events=false
```

同机观测：

| 指标 | 结果 |
|---|---:|
| GPU collector `collector/check_weight_update` mean | 12.336 ms |
| GPU collector median / p95 | 9.994 / 19.683 ms |
| GPU collector `read_weights_into` 稳态范围 | 约 8.4–19.8 ms |
| CPU collector outer median / p95 | 0.585 / 0.766 ms |
| learner `write_weights` mean | 约 0.51 ms |
| 同一 actor、GPU 空闲时完整 read/load mean | 约 0.462 ms |

Trace 中 27/27 个启动后的 collector weight read 与 learner `update_critic` / `update_actor` span 重叠，且没有 read 与 learner `write_weights_d2h` 重叠。现有证据因此不支持“1.098 MiB 原始 H2D 带宽需要 10 ms”，更支持以下组合原因：

1. 31 个同步式小 H2D / state apply 调用造成碎片化；
2. collector 与 learner 是同卡不同进程/CUDA context；
3. runner release collector 后立即执行下一轮 learner updates，collector 的 blocking H2D 在 learner kernels 运行期间等待调度；
4. `read_weights_into(sd)` 后的 `actor.load_state_dict(sd)` 可能是重复 apply。

## Deliverable

在不改变共享 snapshot、version 和 lock 语义的前提下，为 CUDA collector 提供低开销 weight apply 路径，并保留 CPU 参考行为。

候选实现按以下顺序验证，不预先绑定具体 PyTorch 私有 API：

1. 证明并移除 `read_weights_into(sd)` 后不必要的二次 `load_state_dict()`；parameter 和 persistent buffer 必须全部正确更新。
2. 复用预计算 layout 和 staging storage，避免每个周期构造 31 个临时 host array；优先将 flat shared snapshot 合并为一次 H2D，再在 device 侧安全 apply 到 actor state。
3. 若单次 H2D 仍被下一轮 learner compute 阻塞，可在不增加 queue/ack、不改变 tick/release 次数的前提下，利用 collector 已有等待窗口提前应用已经发布的 version，使 weight refresh 不再位于下一次 action inference 的临界路径。
4. 保持 metrics 能区分 version check、实际 snapshot read/apply 和 coordination wait；如果修改 trace 名称，需兼容或同步更新对应测试。

若达到性能门槛必须依赖新的 runner handshake、release/ack lifecycle 或 GPU IPC，则停止本 issue 的实现扩张，以 trace 证据另拆并单独确认后续 issue。

## Contract and correctness requirements

- collector actor 保持常驻其 resolved inference device；正常 forward 不重新构造或上传网络。
- 一次 version publish 对应一个完整、一致的 actor snapshot；collector 不得看到部分新、部分旧的 state。
- parameter、persistent buffer、shape、dtype 和 state name 映射保持一致。
- version 只在完整 publish 后可见；现有跨进程 lock/ownership 语义不得弱化。
- 不改变 learner update 次数、actor publish 频率、`env_steps_per_sync`、replay 数据或训练数学语义。
- CPU collector 默认路径行为不变；MPS/XPU 等非目标 accelerator 至少保留正确 fallback，不新增未经验证的 support claim。

### Performance

在同一台机器、同一 commit、相同 workload 和 warm-up/measurement 协议下，至少 5 次 paired baseline/candidate 运行：

- 只统计 version 实际变化的稳态 refresh event，不把 startup initial load 或 no-op version check 混入。
- GPU collector weight refresh median 相对 baseline 至少下降 70%，并达到不高于 3 ms。
- GPU collector weight refresh p95 不高于 5 ms。
- 报告 learner write、collector refresh、action select、env step、coordination 和完整 tick/cycle 时间，证明成本没有被隐藏到其他 timing bucket。
- 端到端 Steps/s 不退化；若改变 refresh 调度，应证明不是通过串行化 collector 与 learner 换取局部 metric 变好。
- CPU collector weight sync 无显著回退；相同协议下 median 回退不超过 10%。

若目标硬件噪声使绝对阈值不可复现，必须保留 paired raw data、median/p95 与 trace，并在合入前由 maintainer 明确确认替代阈值；不能只以单次最好结果验收。

### Correctness

- 新增靠近 owner boundary 的测试，覆盖 CPU 和 CUDA state apply、31-entry parameter/buffer 完整性、version 不变时 no-op、连续 version 更新以及 concurrent publish/read 不产生 torn snapshot。
- 现有定向测试通过：

  ```bash
  uv run pytest -q \
    tests/ipc/test_shared_weight_sync.py \
    tests/algos/test_offpolicy_worker.py \
    tests/algos/test_offpolicy_logger.py
  ```

- 目标 Motrix FlashSAC 命令完成 smoke run，loss/reward/step 推进无明显异常。
- 最终 PR 前通过 `make test-all`。

## Non-goals

- CUDA IPC、CUDA shared storage、跨进程直接 actor alias、NCCL 或 GPU snapshot 双缓冲。
- 新 collector/learner queue、ack、barrier、runner execution path 或 lifecycle。
- 改变 actor 网络结构、checkpoint payload、state-dict contract 或算法数值语义。
- 改变 env/backend NumPy I/O、Motrix backend contract、replay pipeline 或 GPU-resident replay。
- 为一次 benchmark 把业务规则写入 `scripts/`，或新增常规 CI / 永久 GPU benchmark infrastructure。
- 同时解决 learner tiny-kernel、整体 GPU 利用率或其他 off-policy 性能问题。

## Owner files and validation boundary

主要 owner：

- `src/unilab/ipc/weight_sync.py`
- `src/unilab/algos/torch/offpolicy/worker.py`
- `tests/ipc/test_shared_weight_sync.py`
- `tests/algos/test_offpolicy_worker.py`
- 必要时更新邻近 timing/logger test

本 issue 的性能 claim 只针对上述 CUDA + G1 FlashSAC + Motrix 复现 workload；其他算法/backend 只要求 contract no-regression，不在没有数据时宣称同等性能收益。

## Dependencies and related work

- #687：引入 `training.collector_infer_device`，并将 GPU-aware IPC 明确留作非目标。
- #651：off-policy collector active-window timing 与 throughput 历史。
- #662：learner tiny-kernel/launch-bound 是独立问题，不纳入本 issue。

创建本 issue 只授权上述规划；开始实现仍需 maintainer 单独确认。
